src/vm/eventpipebuffermanager.cpp

```diff
@@ -1050,19 +1050,30 @@ void EventPipeBufferManager::SuspendWriteEvent(uint32_t sessionIndex)
 
     // Wait for any straggler WriteEvent threads that may have already allocated a buffer but
     // hadn't yet relinquished it.
-    {
+    bool hasWriteInProgress = false;
```

Comment:
The change you've got looks like it works and if it is already reasonably tested I would stick with it. Fwiw if this change isn't already tested a simpler alternative could be to delete the entire loop from 1051- 1094 and add one line to the loop above:
```C++
for (size_t i = 0; i < threadList.Size(); i++)
    {
        EventPipeThread *pThread = threadList[i];
        {
            SpinLockHolder _slh(pThread->GetLock());
            EventPipeThreadSessionState *const pSessionState = pThread->GetSessionState(m_pSession);
            pSessionState->SetWriteBuffer(nullptr);
        }
        // Add this line here. We are outside the m_lock so the deadlock we hit below won't occur here.
        YIELD_WHILE(pThread->GetSessionWriteInProgress() == sessionIndex);
    }
```

Reply:
This hasn't been validated beyond passing the existing tests yet (hence draft mode), so I'll look at your change. My only concern on cursory glance is that we are calling `pThread->GetSessionWriteInProgress()` outside the lock and I can't recall if that requires `m_lock` or not. I'll read the code to find out.

Reply:
As I recall this is just reading a volatile field and is designed to be lock-free, but checking is definitely better than relying on my memory : )